File: mod_src/OpenComputers-main-MC1.21.1/src/data/java/li/cil/oc/data/DataGenerators.java
package li.cil.oc.data;

import li.cil.oc.OpenComputers;
import net.minecraft.Util;
import net.minecraft.core.RegistrySetBuilder;
import net.minecraft.core.registries.Registries;
import net.minecraft.data.loot.LootTableProvider;
import net.minecraft.data.registries.RegistryPatchGenerator;
import net.minecraft.world.level.storage.loot.parameters.LootContextParamSets;
import net.neoforged.bus.api.SubscribeEvent;
import net.neoforged.fml.common.EventBusSubscriber;
import net.neoforged.neoforge.common.data.AdvancementProvider;
import net.neoforged.neoforge.common.data.DatapackBuiltinEntriesProvider;
import net.neoforged.neoforge.data.event.GatherDataEvent;

import java.util.List;
import java.util.Set;

@EventBusSubscriber
public class DataGenerators {
    @SubscribeEvent
    public static void gatherData(GatherDataEvent event) {
        var generator = event.getGenerator().getVanillaPack(true);
        var existingFiles = event.getExistingFileHelper();

        var registriesPatch = RegistryPatchGenerator.createLookup(
            event.getLookupProvider(),
            Util.make(new RegistrySetBuilder(), builder -> {
                builder.add(Registries.DAMAGE_TYPE, OCDamageTypeBootstrap::bootstrap);
            })
        );
        var registries = registriesPatch.thenApply(RegistrySetBuilder.PatchedRegistries::full);

        generator.addProvider(o -> new DatapackBuiltinEntriesProvider(o, registriesPatch, Set.of(OpenComputers.ID())));

        var blockTags = generator.addProvider(o -> new OCBlockTagsProvider(o, registries, existingFiles));
        generator.addProvider(o -> new OCItemTagsProvider(o, registries, blockTags.contentsGetter(), existingFiles));
        generator.addProvider(o -> new OCDamageTypeTagsProvider(o, registries, existingFiles));

        generator.addProvider(o -> new OCEEPROMProvider(o, registries, existingFiles));
        generator.addProvider(o -> new OCLootDiskProvider(o, registries, existingFiles));

        generator.addProvider(o -> new OCRecipeProvider(o, registries));
        generator.addProvider(o -> new AdvancementProvider(o, registries, existingFiles, List.of(new OCAdvancementProvider())));

        generator.addProvider(o -> new LootTableProvider(o, Set.of(), List.of(
            new LootTableProvider.SubProviderEntry(OCBlockLoot::new, LootContextParamSets.BLOCK),
            new LootTableProvider.SubProviderEntry(r -> new OCPresentLoot(), LootContextParamSets.EMPTY)
        ), registries));

        generator.addProvider(o -> new OCItemModelProvider(o, existingFiles));
        generator.addProvider(o -> new OCBlockStateProvider(o, existingFiles));
    }
}

File: src/main/scala/li/cil/oc/data/DataGenerators.scala
package li.cil.oc.data

import java.util
import java.util.function.Consumer

import net.minecraft.advancements.AdvancementHolder
import net.minecraft.core.HolderLookup
import net.minecraft.data.advancements.AdvancementProvider
import net.minecraft.data.advancements.AdvancementSubProvider
import net.neoforged.neoforge.data.event.GatherDataEvent
import net.neoforged.bus.api.SubscribeEvent

object DataGenerators {
  @SubscribeEvent
  def gatherData(event: GatherDataEvent): Unit = {
    val generator = event.getGenerator

    generator.addProvider(
      event.includeServer(),
      new AdvancementProvider(
        generator.getPackOutput,
        event.getLookupProvider,
        util.List.of(new AdvancementSubProvider {
          // 1.21.1：AdvancementSubProvider 回调的第二个参数类型是 Consumer[AdvancementHolder]。
          override def generate(registries: HolderLookup.Provider, writer: Consumer[AdvancementHolder]): Unit = {
            Advancements.generate(registries, writer)
          }
        })
      )
    )
  }
}
